libr/debug/p/debug_native.c

```diff
@@ -240,28 +242,34 @@ static int r_debug_native_continue(RDebug *dbg, int pid, int tid, int sig) {
 		r_cons_break_push ((RConsBreak)interrupt_process, dbg);
 	}
 
-	if (dbg->continue_all_threads && dbg->n_threads) {
-		RList *list = dbg->threads;
+	ret = r_debug_ptrace (dbg, PTRACE_CONT, tid, NULL, (r_ptrace_data_t)(size_t)contsig);
+	if (ret == -1) {
+		r_sys_perror ("PTRACE_CONT");
+	}
+
+	// Continue all threads that are stopped by the debugger only
+	// to keep other signals in the queue for the next waitpid
+	// by checking for SIGSTOP signals sent by the debugger
+	if (dbg->continue_all_threads && dbg->threads) {
 		RDebugPid *th;
 		RListIter *it;
-
-		if (list) {
-			r_list_foreach (list, it, th) {
-				if (th->pid) {
-					ret = r_debug_ptrace (dbg, PTRACE_CONT, th->pid, NULL, (r_ptrace_data_t)(size_t)contsig);
-					if (ret) {
-						perror ("PTRACE_CONT");
-					}
+		pid_t dbgpid = getpid ();
+		r_list_foreach (dbg->threads, it, th) {
+			siginfo_t siginfo = { 0 };
+			ret = r_debug_ptrace (dbg, PTRACE_GETSIGINFO, th->pid, 0, (r_ptrace_data_t)(size_t)&siginfo);
+			// Skip if the thread is already running
+			if (ret == -1) {
+				continue;
+			}
+			if (siginfo.si_signo == SIGSTOP && siginfo.si_pid == dbgpid) {
```

Comment:
> This will fail forever for the original thread after switching to a different thread with dpt=. The thread won't resume even if you re-attach to it.

You are right.

> Why not continue if there's another signal? It should be waitpid'd and handled.

I have noticed that even if there is a pending signal in the thread, PTRACE_CONT will continue the thread without errors and the pending signal will be lost after the that thread continues running.

For example when, a thread gets signaled before it is stopped by the debugger's SIGSTOP during linux_stop_threads, the pending signal is in the queue, so continuing all threads even if there is another signal will cause the pending signal to be lost:

1. Thread A creates Thread B.
2. `linux_dbg_wait` breaks and returns `R_DEBUG_REASON_NEW_TID`
2. `r_debug_wait` calls `r_debug_stop` and `linux_stop_threads` which tries to stop all running threads.
2. Thread B exits immediately and Thread A traps into `SIGTRAP` with `PTRACE_EVENT_EXIT` before it gets stopped by `linux_stop_threads`. `linux_stop_threads` will not send another `SIGSTOP` to Thread A because it is stopped.
3.  User continues the process. If we continue all thread even if there is a pending signal in Thread A (`SIGTRAP`), it will continue it without errors. The pending signal in Thread A is lost after it is continued by `PTRACE_CONT`.

Reply:
The new solution simply overwrites the pending signal in the child, isn't there a chance we'll miss a signal if two threads had an event at the same time? I think that you should handle the signal properly instead of just resetting it to SIGSTOP.

Reply:
I will open a new issue to triage this problem, since it requires more than handling the signals properly and also dealing with thread switching. Everything happens in an asynchronous manner. For example, stopping and continuing threads cannot be done to all threads at once and new signals can be sent in between calls. When every thread is not properly tracked, we might accidentally continue a thread that has a pending signal, which causes the signal to be lost and sometimes the debugger freezes.